invoice/src/main/resources/org/killbill/billing/invoice/dao/InvoiceSqlDao.sql.stg

```diff
@@ -95,61 +89,86 @@ getInvoiceByAccountRecordIdAfter() ::= <<
    <defaultOrderBy("")>
 >>
 
-searchByAccountOrInvoiceId(ordering) ::= <<
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (<idField("t.")> = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-UNION
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (t.account_id = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-order by <recordIdField("")> <ordering>
-limit :rowCount offset :offset
-;
->>
-
-getSearchByAccountOrInvoiceIdCount() ::= <<
-select count(*) from
-((select
-<allTableFields("t.")>
-from <tableName()> t
-where (<idField("t.")> = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-UNION
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (t.account_id = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)) as count
-;
+invoiceBalanceQuery() ::= <<
+SELECT
+      x.invoice_id
+    , x.tenant_record_id
+    , SUM(x.amount) as balance
+FROM (
+    SELECT
```

Comment:
Do we need that first query - is it because we want to return also that `DRAFT` or `VOID` invoices?

Reply:
We could:

* Modify `InvoiceModelDao` to have a `balance` field. This is set only in this new codepath, when computing the balance in SQL (the `balance` column from the SQL query should automatically populate it if the getter and setter for balance are added in `InvoiceModelDao`.
* Modify `DefaultInvoice` to have a private `balance` field. If set during construction of the object (pulled from `InvoiceModelDao`), `getBalance()` returns it, otherwise, it goes through the computation like today.

@reshmabidikar Could you give this a try? When doing searches, the invoices will still be shallow, but when `searchInvoicesByBalance` is called, the balance should be returned. If this work, we could also look at updating the other search queries to do the same?

Reply:
Implemented in https://github.com/killbill/killbill/pull/2023. See https://github.com/killbill/killbill/pull/2021#discussion_r1627965080.